### ISSUE TITLE
Omitting parts of the foreign key in onKey(TableField…) doesn't produce the expected result

### ISSUE THREAD

**Issue by AnghelLeonard** [jOOQ/jOOQ#11150]
Expected behavior: Having multiple potential matches, depending on the omitted parts of the foreign key, the onKey(TableField...) should pick up the expected foreign key.

Actual behavior:  Having multiple potential matches, omitting different parts of the foreign key in onKey(TableField...) produces the same result - the same/wrong foreign key is picked up.

Steps to reproduce the problem:

Having two tables, `PRODUCTLINE` and `PRODUCTLINEDETAIL`, and in `PRODUCTLINEDETAIL` there are two foreign keys referencing `PRODUCTLINE` as follows:

```SQL
...
CONSTRAINT productlinedetail_ibfk_1 FOREIGN KEY (product_line,code) 
    REFERENCES productline (product_line,code),
CONSTRAINT productlinedetail_ibfk_2 FOREIGN KEY (product_line) 
    REFERENCES productline (product_line)
```
The following usages of `onKey(TableField...)` produces the same SQL:

```Java
(1) .onKey(PRODUCTLINEDETAIL.PRODUCT_LINE)
(2) .onKey(PRODUCTLINEDETAIL.CODE)
(3) .onKey(PRODUCTLINEDETAIL.PRODUCT_LINE, PRODUCTLINEDETAIL.CODE)
```
And, the SQL is:

```SQL
on (`classicmodels`.`productlinedetail`.`product_line` 
   = `classicmodels`.`productline`.`product_line` 
and `classicmodels`.`productlinedetail`.`code` 
   = `classicmodels`.`productline`.`code`)
```
I was expecting this SQL only for case (3). For case (1) I was expected the following SQL:

```SQL
on `classicmodels`.`productlinedetail`.`product_line` 
  = `classicmodels`.`productline`.`product_line`
```

While for case (2) to throw an exception that foreign key was not found.

This scenario works fine if the generated `Keys.ForeignKey`s are used.

Versions: 

- jOOQ: 3.14.4
- Java: 14
- Database (include vendor): MySQL, PostgreSQL, SQLServer, Oracle
- OS: Windows 10
- JDBC Driver (include name if inofficial driver): latest JDBC drivers for the above databases

**Comment by lukaseder** [comment]
Thanks a lot for your report. That's clearly a bug.

There's a "feature" that allows for passing only parts of the key definition to the `onKey(TableField...)` method, for "convenience". I'm not sure if that was a good idea in the first, place, but clearly, in case there is ambiguity, we should first try to find an *exact* match before considering these "convenient matches".

**Comment by AnghelLeonard** [comment]
Hi Lukas,
    I hope you have a couple of minutes to clarify this for me. I have a converter (attached source code) and I try this:
Integer[] arr = {201212, 202011}; List<Integer> list = List.of(arr);
YearMonth ym = YEARMONTH.convert(arr[0]); // Work, result: 2012-12
// these throw, DataTypeException: Cannot convert from // 201212 (class java.lang.Integer) to class java.time.YearMonthYearMonth[] yms = YEARMONTH.convert(arr);List<YearMonth> yml = YEARMONTH.convert(list);
Just curious ... is this the expected behavior?
Cheers,Anghel
    On Monday, December 21, 2020, 12:15:39 PM GMT+2, Lukas Eder <notifications@github.com> wrote:  
 
 


Thanks a lot for your report. That's clearly a bug.

There's a "feature" that allows for passing only parts of the key definition to the onKey(TableField...) method, for "convenience". I'm not sure if that was a good idea in the first, place, but clearly, in case there is ambiguity, we should first try to find an exact match before considering these "convenient matches".

**Comment by lukaseder** [comment]
I'll reply by email, this isn't really related to this issue...

**Comment by lukaseder** [comment]
I can confirm this issue. In particular, the order of generation of foreign keys is relevant (and it shouldn't be, of course). Change the names of the keys so they're sorted in the inverse order, alphanumerically, and you'll get the expected result.

**Comment by AnghelLeonard** [comment]
Got it! Thanks!

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.5 (#11171). Thanks again for your report!
